### PR TITLE
refactor: move callback init to builder

### PR DESCRIPTION
**Pull request by nshaaban-cPacket**
Refactor to make IllegalActionAfterOpen impossible in engineio. This is not a breaking change thanks to @1c3t3a having the wisdom not to make public yet. Thank you.

### CODE REVIEW

File: engineio/src/callback.rs [hunk 1]
```diff
@@ -0,0 +1,89 @@
+use crate::Packet;
+use bytes::Bytes;
+use std::fmt::Debug;
+use std::ops::Deref;
+use std::sync::Arc;
+
+pub(crate) type Callback<I> = dyn Fn(I) + 'static + Sync + Send;
+
+#[derive(Clone)]
+/// Internal type, only implements debug on fixed set of generics
```
Comment (1c3t3a) [inline engineio/src/callback.rs:10]: That's nice!

File: engineio/src/client/socket.rs [hunk 2]
```diff
@@ -32,19 +38,72 @@ impl SocketBuilder {
             headers: None,
             tls_config: None,
             handshake: None,
+            on_close: OptionalCallback::default(),
```
Comment (1c3t3a) [inline engineio/src/client/socket.rs:41]: Didn't you plan on putting those in a map? Or is that something that happens in a separate pr?

File: engineio/src/client/socket.rs [hunk 3]
```diff
@@ -32,19 +38,72 @@ impl SocketBuilder {
...
+    /// Registers the `on_packet` callback.
+    pub fn on_packet<T>(mut self, callback: T) -> Self
+    where
+        T: Fn(Packet) + 'static + Sync + Send,
+    {
+        self.on_packet = OptionalCallback::new(callback);
+        self
+    }
+
+    /// Preforms the handshake
```
Comment (1c3t3a) [inline engineio/src/client/socket.rs:106]: Performs

File: engineio/src/client/socket.rs (hunk 2 above)
Comment (1c3t3a) [reply engineio/src/client/socket.rs:41]: Ah wait that's the builder, nevermind!

**Review by 1c3t3a — COMMENTED** [review]
LGTM

File: engineio/src/client/socket.rs (hunk 2 above)
Comment (nshaaban-cPacket) [reply engineio/src/client/socket.rs:41]: I attempted to put them both in a map and struct, opted to go for the way it was.
Map issues:
- The generic type varies between the callbacks, could use a dyn object to try to capture this, that has it's own issues
Struct issues:
- No real value add
- In order to retrieve the values from the struct the function name to get the callback out (on_close or close) taking no parameters looks more (from a socketio background) like a function that would call the events or handlers, not simple getters
- Difficult to get type checking on enclosures to resolve correctly due to lack of trait aliases in rust

File: engineio/src/client/socket.rs (hunk 3 above)
Comment (nshaaban-cPacket) [reply engineio/src/client/socket.rs]: 961d1fb3a6f7a45a474966c1cd64ae83118b9faa

File: engineio/src/client/socket.rs (hunk 2 above)
Comment (1c3t3a) [reply engineio/src/client/socket.rs:48]: Ah oke I See! Didn't think about the generic type parameter. The type parameter could be refactored into an enum, but I'm not sure if that brings some enhancement or just makes things more difficult. I think the current way is pretty straight forward.

**Review by 1c3t3a — APPROVED** [review]
Just some merge conflicts and we're ready to go!

**Comment by nshaaban-cPacket** [comment]
Sending this PR into Draft as it(?)'s intermittently failing with "invalid namespace" https://github.com/1c3t3a/rust-socketio/runs/3465528073?check_suite_focus=true https://github.com/1c3t3a/rust-socketio/runs/3465739096?check_suite_focus=true

**Pull request merged**